File: src/main/java/jade/Scene.java
package jade;


import renderer.Renderer;

import java.util.ArrayList;
import java.util.List;

public abstract class Scene {
    protected Camera camera;
    private boolean isRunning = false;
    protected List<GameObject> gameObjects = new ArrayList<>();
    protected Renderer renderer = new Renderer();

    public Scene() {

    }

    public void init() {

    }

    public void start() {
        gameObjects.forEach(go -> {
            go.start();
            renderer.add(go);
        });
        isRunning = true;
    }

    public void addGameObjectToScene(GameObject go) {
        if (!isRunning) {
            gameObjects.add(go);
        } else {
            gameObjects.add(go);
            go.start();
            renderer.add(go);
        }

    }

    public abstract void update(float dt);

    public Camera camera() {
        return this.camera;
    }
}

File: src/main/java/jade/Scene.kt
package jade

import imgui.ImGui
import renderer.Renderer
import java.util.function.Consumer

abstract class Scene {
    protected lateinit var camera: Camera
    protected var activeGameObject: GameObject? = null
    private var isRunning = false
    protected var gameObjects: MutableList<GameObject> = ArrayList()
    protected var renderer: Renderer = Renderer()
    open fun init() {

    }

    fun start() {
        gameObjects.forEach(Consumer { go: GameObject ->
            go.start()
            renderer.add(go)
        })
        isRunning = true
    }

    fun addGameObjectToScene(go: GameObject) {
        if (!isRunning) {
            gameObjects.add(go)
        } else {
            gameObjects.add(go)
            go.start()
            renderer.add(go)
        }
    }

    abstract fun update(dt: Float)
    fun camera(): Camera {
        return camera
    }

    fun sceneImgui() {
        activeGameObject?.let {
            ImGui.begin("Inspector")
            it.imgui()
            ImGui.end()
        }
        imgui()
    }

    open fun imgui() {}
}